### PR TITLE
cs_user.py: Add compatibility for multiple pam modules

### LINKED ISSUE #7313: Which variants of the Python module for pam integration is Cinnamon meant to support?
https://github.com/linuxmint/Cinnamon/blob/2dc04b8a26fa7a672c579ed62a70dca6c6e4b295/files/usr/share/cinnamon/cinnamon-settings/modules/cs_user.py#L3-L6

AFAICT this is meant to support multiple variants of python2-pam, introduced in 768238facbbabed1e7f0182e3e6ccb119c85afdb as nothing but an import alias.

I don't think it makes much sense at all to attempt to support the same module under multiple names, as `import PAM` will never be exactly the same module as `import pam`...
Unless there is another version I'm unaware of -- I only know of the mysterious `PAM` module used in Debian which has no existing upstream, and several forks of https://atlee.ca/software/pam/ which provide python3 support.

I discovered this while trying to clean up the Arch Linux package, which uses the `pam` module from atlee.ca, not the `PAM` module in Debian.

Unfortunately, if that is what this is intended to deal with, it fails due to the fact that they have a very different API. The `PAM` module uses https://github.com/linuxmint/Cinnamon/blob/c71b0c257bc0d664f64b7fbf25622a1135112401/files/usr/share/cinnamon/cinnamon-settings/modules/cs_user.py#L349-L356

Whereas `pam` seems to expect `import pam; pam.authenticate(username, password, 'passwd')`

Obviously this doesn't work for us at all (and never did, I guess), but I thought I would ask here before simply deleting and replacing our python2-pam package. If the intent is to support both, you can check `hasattr(PAM, 'authenticate')` to see if you can use the (IMHO a lot cleaner) atlee.ca version.

### PR DESCRIPTION
**Pull request by eli-schwartz**
Rework the authentication handling to work with either backend module:
- python pam, ctypes bindings created by Chris Atlee, available on PyPI
  and some distros. Forked for python3 support, and available at PyPI as
  python-pam and https://github.com/FirefighterBlu3/python-pam
- python PyPAM, CPython binary extension, available on some other
  distros, and referencing a dead upstream website (Debian hosts and
  builds from *_orig.tar.gz so it is still able to be built).

The current method allows importing either module without failure, but
then spawns internal errors if pam is used instead of PyPAM due to the
two modules containing entirely different APIs.

Also re-raise internal errors and change the tooltip to indicate that
something went wrong. If something unidentifiable happened, error
messages should not be ruthlessly squelched -- this hindered my efforts
to discover why pam was not working on Arch Linux in the first place.

Fixes #7313

### CODE REVIEW

File: files/usr/share/cinnamon/cinnamon-settings/modules/cs_user.py [hunk 1]
```diff
@@ -343,23 +348,38 @@ def _on_new_password_icon_released(self, widget, icon_pos, event):
...
+
+    def auth_PyPAM(self):
+        auth = PAM.pam()
+        auth.start('passwd')
+        auth.set_item(PAM.PAM_USER, GLib.get_user_name())
+        auth.set_item(PAM.PAM_CONV, self.pam_conv)
+        try:
+            auth.authenticate()
+            auth.acct_mgmt()
+            return True
```
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply files/usr/share/cinnamon/cinnamon-settings/modules/cs_user.py]: rebase issue...

**Comment by clefebvre** [comment]
Wait wait wait :)

Sorry for being late to the party here.. do we NEED to support both?

I know it looks like we did, even though, as you pointed out, we really didn't.

Can we ditch pam entirely and simply just use PAM (i.e. python3-pam) going forward?

**Comment by eli-schwartz** [comment]
1) It looks fairly simple to support both, and it is certainly one way of ensuring no one else ever gets confused about which kind of pam module to use.

2) It would mean that distros which provide pam and not PyPAM, do not have to do anything other than update. No need to add new packages or drop old ones. I *could* do this work if I had to, yes, but then again I also don't know what other distros do.

3) Just looking at the code I changed, makes me wonder what on earth the original author of PyPAM was thinking. You could just as easily ask, why not switch to the pam module instead, which has a single function asking yes/no instead of a six-step "conversation" complete with helper functions that only returns useful data through a series of error exceptions... IMHO auth_pam() is far more pythonic.

4) PyPAM 4.2.0's python3 support seems to exist entirely as unresolved patch files floating around the Debian bugtracker, and it itself also exists only via the Debian build infrastructure. Fedora seems to have an unrelated PyPAM 5.0.0 with an entirely different set of patch files e.g. patches for CVEs, memory leaks.
Whereas pam has a Github fork with python3 support written in pure python that does not seem to have any of these issues.

**Comment by clefebvre** [comment]
OK.

Quick typo:

`self.auth_pam() if pam else self.auth_PyPam()`

should be:

`self.auth_pam() if pam else self.auth_PyPAM()`

**Comment by eli-schwartz** [comment]
Ugh, checking for typos with something I don't even use. :D

Force-pushed a typo fix.

**Pull request merged**